File: app/src/main/java/com/lazycouple/restapiclient/data/RestRepository.java
package com.lazycouple.restapiclient.data;


import com.lazycouple.restapiclient.db.model.Api;
import com.lazycouple.restapiclient.db.model.Parameter;

import java.util.List;

import io.realm.Realm;
import rx.Observable;

/**
 * Created by amuyu on 2017. 6. 13..
 */

public class RestRepository implements DataSource {

    private static RestRepository INSTANCE = null;
    private final DataSource localDataSource;

    public RestRepository(DataSource localDataSource) {
        this.localDataSource = localDataSource;
    }

    public static RestRepository getInstance(DataSource localDataSource) {
        if(INSTANCE == null) {
            INSTANCE = new RestRepository(localDataSource);
        }
        return INSTANCE;
    }

    @Override
    public Observable<List<String>> getHistories() {
        return localDataSource.getHistories();
    }

    @Override
    public Observable<List<Api>> getApiHistories(Realm realm) {
        return localDataSource.getApiHistories(realm);
    }

    @Override
    public void addApi(String url, List<Parameter> parameters) {
        localDataSource.addApi(url, parameters);
    }

    @Override
    public Observable<Api> getApi(Realm realm, String id) {
        return localDataSource.getApi(realm,id);
    }

    @Override
    public void clearApiHistories() {
        localDataSource.clearApiHistories();
    }


}

File: app/src/main/kotlin/com/lazycouple/restapiclient/repository/RestRepository.kt
package com.lazycouple.restapiclient.repository


import com.lazycouple.restapiclient.repository.local.model.Api
import com.lazycouple.restapiclient.repository.local.model.Parameter
import com.lazycouple.restapiclient.ui.presenter.RestRequestPresenter
import io.realm.Realm
import rx.Observable

/**
 * Created by amuyu on 2017. 6. 13..
 */

class RestRepository(private val localDataSource: DataSource) : DataSource {

    override fun getHistories(): Observable<List<String>>? {
        return localDataSource.getHistories()
    }

    override fun getApiHistories(realm: Realm): Observable<List<Api>> {
        return localDataSource.getApiHistories(realm)
    }

    override fun addApi(url: String, method: RestRequestPresenter.Method, parameters: List<Parameter>) {
        localDataSource.addApi(url, method, parameters)
    }

    override fun getApi(realm: Realm, id: String): Observable<Api> {
        return localDataSource.getApi(realm, id)
    }

    override fun clearApiHistories() {
        localDataSource.clearApiHistories()
    }

    companion object {
        var INSTANCE: RestRepository? = null

        fun getInstance(localDataSource: DataSource): RestRepository {
            if (INSTANCE == null) {
                INSTANCE = RestRepository(localDataSource)
            }
            return INSTANCE!!
        }
    }


}
